### PR TITLE
Specialize numpy scalar module attrs

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #183994

Treat NumPy floating module attributes as guarded scalar constants so they do not enter compiled graphs as CPU tensor inputs. Preserve NumPy dtype semantics for tensor factories and keep CUDA graphs free of scalar DeviceCopy partitions.\n\nFixes #172942\nGenerated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — COMMENTED** [review]
Would it be better to copy these to pinned memory and convert to cuda ? We do already have a path for handling this for 0-dim tensors in post grad

**Comment by jansel** [comment]
My agent says I don't think pinned-memory conversion is the better fix for this PR. The existing 0-dim path is for compiled autograd/backward scalar tensors; here Dynamo is turning an immutable NumPy module attribute into a CPU tensor input before Inductor, so Inductor still sees a host-side input/DeviceCopy and partitions cudagraphs. Specializing the NumPy scalar in Dynamo as a guarded scalar constant removes that CPU input entirely and preserves dtype behavior for tensor factories. Pinned-copying is still useful for genuinely dynamic CPU scalar tensors, but this module-attr case is safe to guard and burn in.

**Review by eellison — COMMENTED** [review]
The problem is that we're not giving different behavior from without cudagraphs. and we can also partition out what we dont use.

**Comment by jansel** [comment]
My agent says yes, this PR intentionally fixes the Dynamo policy independently of cudagraphs. The host scalar input is created before Inductor sees the graph; for an immutable NumPy nn.Module attribute we can guard on the NumPy scalar type/value and compile the same Python scalar expression a Python float attr would produce, while dynamic CPU scalar tensors still use the existing paths. Partitioning unused CPU work would not remove the used host-to-device scalar in this repro, so burning in the guarded module attr addresses the root cause.

**Review by eellison — COMMENTED** [review]
i'm still unconvinced this is the right policy decision when we can do cpu->gpu pinned memory transfer with good bandwidth and avoid specialization or missing compile. potentially we can augment the partition information of why we did this instead or give an option there.

**Comment by jansel** [comment]
My agent says I agree pinned-memory CPU-to-GPU transfer is useful for genuinely dynamic CPU scalar tensors, but this PR is fixing a narrower Dynamo policy bug: an immutable NumPy nn.Module scalar attr is currently treated differently from an equivalent Python scalar attr and is materialized as a CPU tensor graph input before Inductor. The change specializes only guarded NumPy floating/signed-integer module attrs, removing the host input and DeviceCopy at the root; improved partition diagnostics or pinned transfer for dynamic CPU inputs can be handled separately without replacing this specialization.

File: test/dynamo/test_modules.py
Comment (williamwen42) [inline test/dynamo/test_modules.py:1]: There is some missing test coverage: richcompare, hash (non-nan), graph break/return reconstruction.

**Review by williamwen42 — COMMENTED** [review]
Overall looks good, but may need to rebase and make changes to take into account the recently landed tp_richcompare PR.

File: test/dynamo/test_modules.py
Comment (jansel) [reply test/dynamo/test_modules.py:1]: My agent says I addressed this by adding focused coverage for direct NumPy scalar richcompare dunder calls, non-NaN hash specialization, and graph-break return reconstruction. The richcompare dunder path now routes through the NumPy-backed comparison handling so escaping results preserve np.bool_ semantics or graph-break when they cannot be represented safely.

**Review by eellison — COMMENTED** [review]
deferring to @williamwen42

**Comment by williamwen42** [comment]
Took a closer look and found one behavior regression plus a registry cleanup item.

## BC: NumPy *methods* on a specialized scalar module attr now graph-break

Specializing the attr as `NumpyScalarVariable` means it no longer goes through `NumpyNdarrayVariable`, so any NumPy method on it hits `Unsupported method call` instead of tracing. Minimal repro:

```python
import numpy as np, torch

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.scale = np.float64(2.0)
    def forward(self, x):
        return x * self.scale.astype(np.float32)   # NumPy method on the attr

m, x = Mod(), torch.randn(4)
torch.compile(m, backend="eager", fullgraph=True)(x)
```

- **On current `main`:** compiles and runs (`.astype` traces through the graph).
- **With this PR:** raises `torch._dynamo.exc.Unsupported: Unsupported method call` under `fullgraph=True`. `.mean()`/`.sum()`/`.round()` etc. fail the same way; `.dtype` access also regresses its message from `Unsupported ndarray attribute access` to `Unexpected type in sourceless builder`.

Plain arithmetic (`x * self.scale`), numpy `+`, and `.item()` are unaffected.

### Easy fix: constant-fold scalar methods

Since the scalar is guarded by exact value (`TYPE_MATCH` + `CONSTANT_MATCH`), scalar-returning methods are pure functions of the constant — same justification already used for `.item()`/`.real`/`.imag`. ~20 lines in `NumpyScalarVariable.call_method`, after the existing `item` handling:

```python
member = getattr(type(self.numpy_value), name, None)
if callable(member):
    try:
        const_args = [a.as_python_constant() for a in args]
        const_kwargs = {k: v.as_python_constant() for k, v in kwargs.items()}
    except NotImplementedError:
        return super().call_method(tx, name, args, kwargs)
    try:
        result = getattr(self.numpy_value, name)(*const_args, **const_kwargs)
    except (TypeError, ValueError) as exc:
        raise_observed_exception(type(exc), tx, args=list(exc.args))
    numpy_result = self.maybe_create_supported(result, f"{self.numpy_type.__name__}.{name}")
    if numpy_result is not None:
        return numpy_result
    if ConstantVariable.is_literal(result):
        return ConstantVariable.create(result)
return super().call_method(tx, name, args, kwargs)
```

Verified (PR + this snippet, `fullgraph=True`): `.astype(np.float32)`, `.mean()`, `.sum()`, `.round()` all compile and match eager (incl. dtype), the value is still baked in (graph has only `x` as a non-symint input, no `self_scale`), and it still recompiles when the attr value changes.

Limitation worth noting: array-returning methods (e.g. `self.scale.reshape((1,)).sum()`) still graph-break cleanly — folding only covers scalar/literal results. Full pre-PR parity for those would require lazily re-wrapping via the source as a `NumpyNdarrayVariable`, which re-introduces the CPU input the PR removes, so it's only worth it for genuine array work. Either way, please route the unsupported case through an explicit `unimplemented()` with a real `gb_type` rather than the current `NotImplementedError` path, and add a test pinning both behaviors (scalar method folds; array method graph-breaks).

## Graph break registry: stable GBID got repurposed

(Correcting my earlier note here — the registry can be hand-edited; the rules are just never delete entries and append edits as new entries under the *same* GBID. So `GB7294` having two `numpy_scalar_comparison_result` entries that differ only by `Context` is fine/expected.)

The real issue is `GB0225`. On `main` it's the stable, single-entry ID for `Gb_type: "Attempted to use tensor creation function with requires_grad=True"`. This PR:

- inserts the brand-new gb_type `numpy_scalar_tensor_requires_grad` as a new entry **under `GB0225`**, and
- mints a new `GB5225` carrying a duplicate of `GB0225`'s original `"Attempted to use tensor creation function with requires_grad=True"`.

So a new gb_type was added under an existing unrelated GBID, and the long-standing `"Attempted to use..."` entry now also shows up under a fresh `GB5225` — effectively moving/duplicating a stable ID, which breaks GBID stability for tlparse. Per the append-only rule, the new `numpy_scalar_tensor_requires_grad` should get its own new GBID and `GB0225` should keep only its original entry (no deletions of existing entries). Easiest to let `tools/dynamo/gb_id_mapping.py` assign the new ID rather than hand-placing it.

**Comment by jansel** [comment]
My agent says I addressed the NumPy scalar method regression and graph-break registry feedback. The update constant-folds only allowlisted pure scalar methods (astype/mean/sum/round/tolist), graph-breaks unsupported scalar methods before invoking them, restores GB0225 to the original requires_grad=True gb_type, and assigns separate GBIDs for the NumPy-specific graph breaks. I also added focused tests for scalar method folding, array-returning method graph breaks, and np.longlong guard support.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the failed Inductor TorchBench training job failed during the forced-eager reference comparison (eager_two_runs_differ for mnasnet1_0/mobilenet_v2) before Dynamo produced any graphs, so it does not appear caused by the NumPy scalar tracing change in this PR. I rebased onto current main and resubmitted the stack to restart CI on a fresh head.

**Comment by jansel** [comment]
My agent says I addressed the remaining NumPy scalar `.dtype` concern while rebasing onto current `main`. The scalar tracker now uses the live `getattro_impl` protocol, delegates callable attributes through the existing method path, and represents dtype results with `NumpyDTypeVariable`; new coverage verifies both direct `.dtype` and a float64-to-float32 `astype(...).dtype` under `fullgraph=True`. I also resolved the rebase conflict by retaining current-main builtin behavior and reapplying only the NumPy scalar comparison handling. All 18 focused tests, their 18 dynamic-shape variants, the CUDA cudagraph regression, the unspecialized-scalar test, and `lintrunner -a` pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
